File: app/src/main/java/com/example/permissions/MainActivity.java
package com.example.permissions;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.widget.TextView;
import android.widget.Toast;

import java.util.Arrays;

public class MainActivity extends AppCompatActivity {

    static final private String TAG = MainActivity.class.getSimpleName();
    private static final int PERMISSION_STATE = 0;
    private Boolean callPermissionSettings,requestGranted, neverAskAgain;
    private TextView txtData;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        txtData = (TextView) findViewById(R.id.txtData);
    }

    @Override
    protected void onPause() {
        super.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (checkPermission()) requestPermissions();
    }

    private boolean checkPermission() {
        if (requestGranted == null || neverAskAgain == null) {
            return true;
        }
        else if (!requestGranted && !neverAskAgain) {
            txtData.setText("All permission denied");
            return true;
        }
        else if (neverAskAgain && !requestGranted) {
            txtData.setText("All permission denied and set to Don't Ask Again");
            return false;
        }
        else {
            txtData.setText("All permission allowed");
            return false;
        }
    }

    private boolean checkAppPermissionSettings() {
        if (!callPermissionSettings && neverAskAgain && !requestGranted) {
            callPermissionSettings = true;
            return true;
        } else {
            return false;
        }
    }

    private void requestPermissions() {
        Log.d("PermissionsResult", "requestPermissions()");
        String[] manifestPermissions;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
            manifestPermissions = new String[]{
                    Manifest.permission.READ_PHONE_STATE,
                    Manifest.permission.READ_CONTACTS,
                    Manifest.permission.WRITE_CONTACTS,
                    Manifest.permission.READ_EXTERNAL_STORAGE,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE
            };
        } else {
            manifestPermissions = new String[]{
                    Manifest.permission.READ_PHONE_STATE,
                    Manifest.permission.READ_CONTACTS,
                    Manifest.permission.WRITE_CONTACTS,
                    Manifest.permission.WRITE_EXTERNAL_STORAGE
            };
        }

        ActivityCompat.requestPermissions(
                this,
                manifestPermissions,
                PERMISSION_STATE
        );
    }

    private boolean isPermissionsGranted(String permission) {
        return ActivityCompat
                .checkSelfPermission(
                        getApplication(),
                        permission
                ) == PackageManager.PERMISSION_GRANTED;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        //region Logs for onRequestPermissionsResult parameters
        Log.d("PermissionsResult", "requestCode " + requestCode);
        Log.d("PermissionsResult", "permissions " + Arrays.toString(permissions));
        Log.d("PermissionsResult", "grantResults " + Arrays.toString(grantResults));
        //endregion
        //region Code for requestCode and grantResults
        requestGranted = true;
        if (this.PERMISSION_STATE == requestCode) {
            for (int grantResult : grantResults) {
                switch (grantResult) {
                    case PackageManager.PERMISSION_GRANTED:
                        Log.d("PermissionsResult", "grantResult Allowed " + grantResult);
                        //requestGranted = true;
                        break;
                    case PackageManager.PERMISSION_DENIED:
                        Log.d("PermissionsResult", "grantResult Denied " + grantResult);
                        requestGranted = false;
                        break;
                    default:
                        requestGranted = false;
                        break;
                }
            }
        }
        //endregion
        //region Code for permissions
        neverAskAgain = true;
        for(String permission : permissions) {
            if(ActivityCompat.shouldShowRequestPermissionRationale(this, permission)) {
                Log.d("PermissionsResult", "permission Denied " + permission);
                neverAskAgain = false;
            } else {
                if(isPermissionsGranted(permission)) {
                    Log.d("PermissionsResult", "permission Allowed " + permission);
                    neverAskAgain = false;
                } else{
                    Log.d("PermissionsResult", "permission set to never ask again " + permission);
                    //neverAskAgain = true;
                }
            }
        }
        //endregion
        //region Code for showAppPermissionSettings
        callPermissionSettings = false;
        if (checkAppPermissionSettings()) showAppPermissionSettings();
        //endregion
        //region Logs for Booleans of requestGranted, neverAskAgain, callPermissionSettings
        Log.d("PermissionsResult", "requestGranted " + requestGranted);
        Log.d("PermissionsResult", "neverAskAgain " + neverAskAgain);
        Log.d("PermissionsResult", "callPermissionSettings " + callPermissionSettings);
        //endregion
    }

    private void showAppPermissionSettings() {
        Log.d("PermissionsResult", "showAppPermissionSettings()");
        Toast.makeText(this, "Hardware Permissions Disabled", Toast.LENGTH_LONG).show();
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                Uri.fromParts("package", getPackageName(), null));
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.addFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
        intent.addFlags(Intent.FLAG_ACTIVITY_EXCLUDE_FROM_RECENTS);
        startActivity(intent);
    }

    private void  onActivityResult() { Log.d(TAG,"onActivityResult()");
        ActivityResultLauncher<Intent> mStartForResult = registerForActivityResult (
                new ActivityResultContracts.StartActivityForResult(), new ActivityResultCallback<ActivityResult>() {
                @Override
                public void onActivityResult(ActivityResult activityResult) {
                    if (activityResult.getResultCode() == Activity.RESULT_OK) {
                        Intent intent = activityResult.getData();   // Handle the Intent
                    }
                    if (activityResult.getResultCode() == ManifestPermission.SETTINGS_PERMISSION_CODE)
                        Toast.makeText(getBaseContext(), "PERMISSION_SETTINGS_CODE",Toast.LENGTH_SHORT).show();
                }
            }
        );
        //mStartForResult.launch();

        ActivityResultLauncher<String> mGetContent = registerForActivityResult (
            new ActivityResultContracts.GetContent(), new ActivityResultCallback<Uri>() {
                @Override
                public void onActivityResult(Uri uri) {
                    // Handle the returned Uri
                }
            }
        );
        //mGetContent.launch();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        Log.d(TAG,"onActivityResult(" + requestCode + ", " + resultCode + ", " + data );
        if (requestCode == ManifestPermission.SETTINGS_PERMISSION_CODE)
            Toast.makeText(this,"PERMISSION_SETTINGS_CODE",Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/permissions/MainActivity.kt
package com.example.permissions

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.example.permissions.databinding.MainBinder

class MainActivity : AppCompatActivity(), View.OnClickListener {

    companion object {
        private val TAG : String = MainActivity::class.java.getSimpleName()
    }

    private var binder : MainBinder? = null

    override fun onCreate(savedInstanceState : Bundle?) {
        binder = DataBindingUtil.setContentView(this@MainActivity, R.layout.activity_main)
        binder?.setLifecycleOwner(this@MainActivity)
        super.onCreate(savedInstanceState)
        //region Check Permission On Click Listener
        binder?.allCheckPermissions?.setOnClickListener(this@MainActivity)
        binder?.telephonyCheckPermissions?.setOnClickListener(this@MainActivity)
        binder?.microphoneCheckPermissions?.setOnClickListener(this@MainActivity)
        binder?.cameraCheckPermissions?.setOnClickListener(this@MainActivity)
        binder?.videoCallCheckPermissions?.setOnClickListener(this@MainActivity)
        binder?.galleryCheckPermissions?.setOnClickListener(this@MainActivity)
        binder?.contactCheckPermissions?.setOnClickListener(this@MainActivity)
        //endregion
        //region Request Permission On Click Listener
        binder?.allRequestPermissions?.setOnClickListener(this@MainActivity)
        binder?.telephonyRequestPermissions?.setOnClickListener(this@MainActivity)
        binder?.microphoneRequestPermissions?.setOnClickListener(this@MainActivity)
        binder?.cameraRequestPermissions?.setOnClickListener(this@MainActivity)
        binder?.videoCallRequestPermissions?.setOnClickListener(this@MainActivity)
        binder?.galleryRequestPermissions?.setOnClickListener(this@MainActivity)
        binder?.contactRequestPermissions?.setOnClickListener(this@MainActivity)
        //endregion
        binder?.showRationalDialog?.setOnClickListener(this@MainActivity)
    }

    override fun onPause() {
        super.onPause()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onClick(view : View) {
        when(view) {
            binder?.allCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.allPermissions,
                    isGranted = { setTextData("All Permissions Granted!") },
                    isDenied = { setTextData("All Permissions Denied!") }
                )
            }
            binder?.telephonyCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.telephonyPermissions,
                    isGranted = { setTextData("Telephony Permissions Granted!") },
                    isDenied = { setTextData("Telephony Permissions Denied!") }
                )
            }
            binder?.microphoneCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.microphonePermission,
                    isGranted = { setTextData("Microphone Permissions Granted!") },
                    isDenied = { setTextData("Microphone Permissions Denied!") }
                )
            }
            binder?.cameraCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.cameraPermission,
                    isGranted = { setTextData("Camera Permissions Granted!") },
                    isDenied = { setTextData("Camera Permissions Denied!") }
                )
            }
            binder?.videoCallCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.videoCallPermission,
                    isGranted = { setTextData("Video Call Permissions Granted!") },
                    isDenied = { setTextData("Video Call Permissions Denied!") }
                )
            }
            binder?.galleryCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.galleryPermissions,
                    isGranted = { setTextData("Gallery Permissions Granted!") },
                    isDenied = { setTextData("Gallery Permissions Denied!") }
                )
            }
            binder?.contactCheckPermissions -> {
                ManifestPermission.checkSelfPermission(this@MainActivity,
                    ManifestPermission.contactPermission,
                    isGranted = { setTextData("Contact Permissions Granted!") },
                    isDenied = { setTextData("Contact Permissions Denied!") }
                )
            }
            binder?.allRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.allPermissions,
                    ManifestPermission.ALL_PERMISSION_CODE
                )
            }
            binder?.telephonyRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.telephonyPermissions,
                    ManifestPermission.TELEPHONY_PERMISSION_CODE
                )
            }
            binder?.microphoneRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.microphonePermission,
                    ManifestPermission.MICROPHONE_PERMISSION_CODE
                )
            }
            binder?.cameraRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.cameraPermission,
                    ManifestPermission.CAMERA_PERMISSION_CODE
                )
            }
            binder?.videoCallRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.videoCallPermission,
                    ManifestPermission.VIDEO_CALL_PERMISSION_CODE
                )
            }
            binder?.galleryRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.galleryPermissions,
                    ManifestPermission.GALLERY_PERMISSION_CODE
                )
            }
            binder?.contactRequestPermissions -> {
                ManifestPermission.requestPermissions(this@MainActivity,
                    ManifestPermission.contactPermission,
                    ManifestPermission.CONTACT_PERMISSION_CODE
                )
            }
            binder?.showRationalDialog -> {
                ManifestPermission.showRationaleDialog(this@MainActivity,"Go to App Permission Settings?", activityResultLauncher)
            }
        }
    }

    private fun setTextData(text : String) {
        binder?.textData?.setText(text)
    }

    override fun onRequestPermissionsResult(requestCode : Int, permissions : Array<String>, grantResults : IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        Log.d("PermissionsResult", "requestCode $requestCode")
        Log.d("PermissionsResult","permissions " + permissions.contentToString())
        Log.d("PermissionsResult","grantResults " + grantResults.contentToString())

        permissions.map { permission ->
            ManifestPermission.checkNeverAskAgain(this@MainActivity,permission,
                isNeverAskAgain = {
                    Log.d("PermissionsResult", "checkNeverAskAgain isNeverAskAgain single $permission")
                },
                isNotNeverAskAgain = {
                    Log.d("PermissionsResult", "checkNeverAskAgain isNotNeverAskAgain single $permission")
                }
            )

            ManifestPermission.checkPermissionsResult(this@MainActivity,permission,
                isNeverAskAgain = {
                    Log.d("PermissionsResult", "checkPermissionsResult isNeverAskAgain single $permission")
                }, isDenied = {
                    Log.d("PermissionsResult","checkPermissionsResult isDenied single $permission")
                }, isGranted = {
                    Log.d("PermissionsResult","checkPermissionsResult isGranted single $permission")
                }
            )
        }

        ManifestPermission.checkNeverAskAgain(this@MainActivity, permissions,
            isNeverAskAgain = {
                Log.d("PermissionsResult", "checkNeverAskAgain isNeverAskAgain")
            },
            isNotNeverAskAgain = {
                Log.d("PermissionsResult", "checkNeverAskAgain isNotNeverAskAgain")
            }
        )

        ManifestPermission.checkPermissionsResult(this@MainActivity, permissions, grantResults,
            isNeverAskAgain =  {
                Log.d("PermissionsResult", "checkPermissionsResult isNeverAskAgain")
                when(requestCode) {
                    ManifestPermission.ALL_PERMISSION_CODE -> { setTextData("All Permissions is Never Ask Again!") }
                    ManifestPermission.TELEPHONY_PERMISSION_CODE -> { setTextData("Telephony Permissions is Never Ask Again!") }
                    ManifestPermission.MICROPHONE_PERMISSION_CODE -> { setTextData("Microphone Permissions is Never Ask Again!") }
                    ManifestPermission.CAMERA_PERMISSION_CODE -> { setTextData("Camera Permissions is Never Ask Again!") }
                    ManifestPermission.VIDEO_CALL_PERMISSION_CODE -> { setTextData("Video Call Permissions is Never Ask Again!") }
                    ManifestPermission.GALLERY_PERMISSION_CODE -> { setTextData("Gallery Permissions is Never Ask Again!") }
                    ManifestPermission.CONTACT_PERMISSION_CODE -> { setTextData("Contact Permissions is Never Ask Again!") }
                }
            }, isDenied = {
                Log.d("PermissionsResult","checkPermissionsResult isDenied")
                when(requestCode) {
                    ManifestPermission.ALL_PERMISSION_CODE -> { setTextData("All Permissions Denied!") }
                    ManifestPermission.TELEPHONY_PERMISSION_CODE -> { setTextData("Telephony Permissions Denied!") }
                    ManifestPermission.MICROPHONE_PERMISSION_CODE -> { setTextData("Microphone Permissions Denied!") }
                    ManifestPermission.CAMERA_PERMISSION_CODE -> { setTextData("Camera Permissions Denied!") }
                    ManifestPermission.VIDEO_CALL_PERMISSION_CODE -> { setTextData("Video Call Permissions Denied!") }
                    ManifestPermission.GALLERY_PERMISSION_CODE -> { setTextData("Gallery Permissions Denied!") }
                    ManifestPermission.CONTACT_PERMISSION_CODE -> { setTextData("Contact Permissions Denied!") }
                }
            }, isGranted = {
                Log.d("PermissionsResult","checkPermissionsResult isGranted")
                when(requestCode) {
                    ManifestPermission.ALL_PERMISSION_CODE -> { setTextData("All Permissions Granted!") }
                    ManifestPermission.TELEPHONY_PERMISSION_CODE -> { setTextData("Telephony Permissions Granted!") }
                    ManifestPermission.MICROPHONE_PERMISSION_CODE -> { setTextData("Microphone Permissions Granted!") }
                    ManifestPermission.CAMERA_PERMISSION_CODE -> { setTextData("Camera Permissions Granted!") }
                    ManifestPermission.VIDEO_CALL_PERMISSION_CODE -> { setTextData("Video Call Permissions Granted!") }
                    ManifestPermission.GALLERY_PERMISSION_CODE -> { setTextData("Gallery Permissions Granted!") }
                    ManifestPermission.CONTACT_PERMISSION_CODE -> { setTextData("Contact Permissions Granted!") }
                }
            }
        )
        ManifestPermission.checkPermissionsResult(this@MainActivity, ManifestPermission.ALL_PERMISSION_CODE, requestCode, permissions, grantResults,
        isNeverAskAgain = {
            Log.d("PermissionsResult","checkPermissionsResult with Selected Request Code, isNeverAskAgain")
        }, isDenied = {
            Log.d("PermissionsResult","checkPermissionsResult with Selected Request Code, isDenied")
        }, isGranted = {
            Log.d("PermissionsResult","checkPermissionsResult with Selected Request Code, isGranted")
        } )
    }

    private val activityResultLauncher : ActivityResultLauncher<Intent> = registerForActivityResult( ActivityResultContracts.StartActivityForResult() ) { activityResult ->
        Log.d(TAG,"activityResultLauncher(${activityResult.getResultCode()},${activityResult.getData()})")
        if (activityResult.getResultCode() == Activity.RESULT_OK) {
            Log.d(TAG,"activityResult.getResultCode() == Activity.RESULT_OK")
        } else if (activityResult.getResultCode() == Activity.RESULT_CANCELED) {
            Log.d(TAG,"activityResult.getResultCode() == Activity.RESULT_CANCELED")
        }
    }

    val requestPermissionLauncher : ActivityResultLauncher<String> =
        registerForActivityResult(ActivityResultContracts.RequestPermission()) {
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        Log.d(TAG,"onActivityResult($requestCode,$resultCode,$data)")
        if (requestCode == ManifestPermission.SETTINGS_PERMISSION_CODE)
            Toast.makeText(this@MainActivity,"PERMISSION_SETTINGS_CODE",Toast.LENGTH_SHORT).show()
    }

    override fun onDestroy() {
        Log.d(TAG,"onDestroy()")
        super.onDestroy()
    }
}